src/persist/clients.ts

```diff
@@ -72,83 +78,151 @@ export async function getClient(
+        bootstrapClient.heartbeatTimestampMs < client.heartbeatTimestampMs
+      ) {
+        bootstrapClient = client;
+      }
     }
-  }
 
-  let newClientCommit;
-  if (bootstrapClient) {
-    const bootstrapCommit = await db.baseSnapshot(
-      bootstrapClient.headHash,
+    const newClientCommitData = await dagStore.withRead(async dagRead => {
+      if (bootstrapClient) {
+        const bootstrapCommit = await db.baseSnapshot(
+          bootstrapClient.headHash,
+          dagRead,
+        );
+        // Copy the snapshot with one change: set last mutation id to 0.  Replicache
+        // server implementations expect new client ids to start with last mutation id 0.
+        // If a server sees a new client id with a non-0 last mutation id, it may conclude
+        // this is a very old client whose state has been garbage collected on the server.
+        return newSnapshotCommitData(
+          bootstrapCommit.meta.basisHash,
+          0 /* lastMutationID */,
+          bootstrapCommit.meta.cookieJSON,
+          bootstrapCommit.valueHash,
+          bootstrapCommit.indexes,
+        );
+      } else {
+        // No existing snapshot to bootstrap from. Create empty snapshot.
+        // TODO This empty btree write is a problem...
+        //const emptyBTreeHash = await new BTreeWrite(dagWrite).flush();
+        return newSnapshotCommitData(
+          null /* basisHash */,
+          0 /* lastMutationID */,
+          null /* cookie */,
+          'emptybtreehash' as unknown as Hash,
+          [] /* indexes */,
+        );
+      }
+    });
+
+    const newClientCommitChunk = dag.createChunkWithHash(
+      await nativeHashOf(newClientCommitData),
+      newClientCommitData,
+      getRefs(newClientCommitData),
+    );
+    return {
+      clients: new Map(clients).set(newClientID, {
+        heartbeatTimestampMs: Date.now(),
+        headHash: newClientCommitChunk.hash,
+      }),
+      chunksToPut: [newClientCommitChunk],
+    };
+  }, dagStore);
+  const newClient = updatedClients.get(newClientID);
+  assertNotUndefined(newClient);
+  return [newClientID, newClient];
+}
+
+function nativeHashOfClients(clients: ClientMap): Promise<Hash> {
+  const data = clientMapToChunkDataNoHashValidation(clients);
+  return nativeHashOf(data);
+}
+
+export const noUpdates = Symbol();
+export type NoUpdates = typeof noUpdates;
+
+export type ClientsUpdate = (
+  clients: ClientMap,
+) => Promise<
+  {clients: ClientMap; chunksToPut?: Iterable<dag.Chunk>} | NoUpdates
+>;
+
+export async function updateClients(
+  update: ClientsUpdate,
+  dagStore: dag.Store,
+): Promise<ClientMap> {
+  const [clients, clientsHash] = await dagStore.withRead(async dagRead => {
+    const clients = await getClients(dagRead);
+    const clientsHead = await dagRead.getHead(CLIENTS_HEAD);
+    return [clients, clientsHead];
+  });
+  return _updateClients(update, clients, clientsHash, dagStore);
+}
+
+export async function _updateClients(
```

Comment:
I don't like the name here ... maybe just updateClientsInternal?

Reply:
done